lib/workflow/task-graph.js

```diff
@@ -0,0 +1,493 @@
+                } else {
+                    return store.getTaskDefinition(taskData.taskName)
+                    .then(function(taskDefinition) {
+                        self._validateTaskDefinition(taskDefinition);
+                        self._validateProperties(taskDefinition, context);
+                        self._validateOptions(taskDefinition, taskData.label);
+                    });
+                }
+            });
+        })
+        .spread(function() {
+            return self;
+        });
+    };
+
+    TaskGraph.prototype._validateTaskLabels = function() {
+        _.transform(this.definition.tasks, function(result, task) {
+            if (result[task.label]) {
+                throw new Error(("The task label '%s' is used more than once in " +
+                                "the graph definition.").format(task.label));
+            } else {
+                result[task.label] = true;
+            }
+        }, {});
+    };
+
+    TaskGraph.prototype._validateTaskDefinition = function(taskDefinition) {
+        assert.object(taskDefinition, 'taskDefinition');
+        assert.string(taskDefinition.friendlyName, 'friendlyName');
+        assert.string(taskDefinition.injectableName, 'injectableName');
+        assert.string(taskDefinition.implementsTask, 'implementsTask');
+        assert.object(taskDefinition.options, 'options');
+        assert.object(taskDefinition.properties, 'properties');
+
+        var baseTaskDefinition = this._getBaseTask(taskDefinition);
+        assert.string(baseTaskDefinition.friendlyName, 'friendlyName');
+        assert.string(baseTaskDefinition.injectableName, 'injectableName');
+        assert.string(baseTaskDefinition.runJob, 'runJob');
+        assert.object(baseTaskDefinition.requiredOptions, 'requiredOptions');
+        assert.object(baseTaskDefinition.requiredProperties, 'requiredProperties');
+        assert.object(baseTaskDefinition.properties, 'properties');
+    };
+
+    TaskGraph.prototype._validateProperties = function(taskDefinition, context) {
+        var self = this;
+        var baseTaskDefinition = self._getBaseTask(taskDefinition);
+        var requiredProperties = baseTaskDefinition.requiredProperties;
+        _.forEach(requiredProperties, function(v, k) {
+            self.compareNestedProperties(
+                v, k, context.properties, baseTaskDefinition.injectableName);
+        });
+
+        // Update shared context with properties from this task
+        var _properties = _.merge(taskDefinition.properties, baseTaskDefinition.properties);
+        context.properties = _.merge(_properties, context.properties);
+    };
+
+    TaskGraph.prototype._validateOptions = function(taskDefinition, label) {
+        var self = this;
+        var baseTaskDefinition = self._getBaseTask(taskDefinition);
+        _.forEach(baseTaskDefinition.requiredOptions, function(k) {
+            var option = taskDefinition.options[k];
```

Comment:
In this case we're inspecting the current options object for the keys specified in the requiredOptions object.